docs/tests.md

```diff
@@ -0,0 +1,454 @@
+correctness. Our KAT testing framework uses this approach to ensure that any
+modifications to the library do not disrupt existing functionality.
+
+Importantly, while pre-calculated parameters and values are common, some
+researchers provide source code for their implementations, serving as **KAT
+generators**. These generators allow for dynamic KAT value generation.
+
+Unlike isolated doctests, KATs take a broader approach by comparing outputs
+across entire estimators, often involving complex and computationally intensive
+operations. To optimize this process, we serialize expected outputs for specific
+predefined inputs. These serialized results streamline subsequent test runs,
+enabling efficient comparisons and saving valuable development time.
+
+Here's a breakdown of how our KAT testing process works:
+
+```mermaid
+flowchart LR;
+  katgen("`KAT Generators`")
+  inputs("`Inputs`")
+  preeo("`Precompute expected outputs`");
+  ser("`Serialize and store inputs/expected outputs`")
+  estimator("`Library estimation`")
+  oc("`TEST:
+    Outputs comparison`")
+
+  katgen --> preeo;
+  inputs --> preeo;
+  preeo --> ser;
+  ser -- inputs --> estimator;
+  estimator -- actual output --> oc;
+  ser -- expected output --> oc;
+```
+
+1. **KAT Generator Functions:** We utilize KAT generator functions located in
+   the `tests/external_estimators` directory. These functions have hardcoded
+   inputs as they are not intended for regular modification. Both input
+   parameters and their corresponding outputs (referred to as `expected_outputs`
+   as they act as reference values) are serialized into the `tests/kat.yaml`
+   file for later use during estimator testing.
+
+2. **Internal Estimation Functions:** With a collection of inputs and their
+   expected outputs, we define how these inputs should be processed within our
+   library. This is achieved through internal estimation functions found in the
+   `tests/internal_estimators` directory. Each function corresponds to a KAT
+   generator in `tests/external_estimators`.
+
+3. **Test Execution and Comparison:** In the final step, we execute all our
+   internal estimation functions using the serialized inputs from
+   `test/kat.yaml`. The calculated outputs are then compared against the
+   expected outputs from the KAT generators to verify the correctness and
+   accuracy of our library's implementations.
+
+This separation of concerns makes it easier to maintain and extend the test
+suite as we introduce new estimators or modify existing ones.
+
+### Writing Tests: A Step-by-Step Guide
+
+This guide walks you through incorporating a new KAT test for a library
+estimator, offering insights relevant for both understanding and updating
+existing tests.
+
+#### 1. Define Your KAT Generator Function
+
+This step involves defining or declaring the KAT generator functions which
+produces the expected outputs from hardcoded inputs. These functions require
```

Comment:
What about this?

"These functions require careful crafting and may originate from external sources or researchers." --> "These functions are wrappers of external estimators of the complexity of particular algorithms. Hence, these functions require careful crafting."

Reply:
I like it. Applied.